src/sinks/util/service/net/tcp.rs

```diff
@@ -0,0 +1,272 @@
+}
+
+/// `TcpConnector` configuration.
+#[configurable_component]
+#[derive(Clone, Debug)]
+pub struct TcpConnectorConfig {
+    /// The address to connect to.
+    ///
+    /// Both IP addresses and hostnames/fully-qualified domain names are accepted formats.
+    ///
+    /// The address _must_ include a port.
+    address: HostAndPort,
+
+    #[configurable(derived)]
+    keepalive: Option<TcpKeepaliveConfig>,
+
+    /// The size of the socket's send buffer, in bytes.
+    ///
+    /// If set, the value of the setting is passed via the `SO_SNDBUF` option.
+    #[configurable(metadata(docs::type_unit = "bytes"))]
+    #[configurable(metadata(docs::examples = 65536))]
+    send_buffer_size: Option<u32>,
+}
+
+impl TcpConnectorConfig {
+    pub const fn from_address(host: String, port: u16) -> Self {
+        Self {
+            address: HostAndPort { host, port },
+            keepalive: None,
+            send_buffer_size: None,
+        }
+    }
+
+    pub const fn set_keepalive(mut self, keepalive: TcpKeepaliveConfig) -> Self {
+        self.keepalive = Some(keepalive);
+        self
+    }
+
+    pub fn as_connector(&self) -> TcpConnector {
+        TcpConnector {
+            address: self.address.clone(),
+            keepalive: self.keepalive,
+            send_buffer_size: self.send_buffer_size,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct TcpConnector {
+    address: HostAndPort,
+    keepalive: Option<TcpKeepaliveConfig>,
+    send_buffer_size: Option<u32>,
+}
+
+impl TcpConnector {
+    async fn connect(&self) -> Result<(SocketAddr, TcpStream), TcpError> {
+        let ip = dns::Resolver
+            .lookup_ip(self.address.host.clone())
+            .await
+            .context(FailedToResolveSnafu)?
+            .next()
+            .ok_or(TcpError::NoAddresses)?;
+
+        let addr = SocketAddr::new(ip, self.address.port);
+
+        let socket = if addr.is_ipv4() {
+            TcpSocket::new_v4().context(FailedToConfigureSnafu)?
+        } else {
+            TcpSocket::new_v6().context(FailedToConfigureSnafu)?
+        };
+
+        if let Some(send_buffer_size) = self.send_buffer_size {
+            if let Err(error) = socket.set_send_buffer_size(send_buffer_size) {
+                warn!(%error, "Failed configuring send buffer size on TCP socket.");
+            }
+        }
+
+        let stream = socket.connect(addr).await.context(FailedToConnectSnafu)?;
+
+        let maybe_keepalive_secs = self
+            .keepalive
+            .as_ref()
+            .and_then(|config| config.time_secs.map(Duration::from_secs));
+        if let Some(keepalive_secs) = maybe_keepalive_secs {
+            if let Err(error) = net::set_keepalive(&stream, keepalive_secs) {
+                warn!(%error, "Failed configuring keepalive on TCP socket.");
+            }
+        }
+
+        Ok((addr, stream))
+    }
+
+    async fn connect_backoff(&self) -> TcpStream {
+        // TODO: Make this configurable.
```

Comment:
just checking- something you intended to do in this PR or a real TODO for laters.

Reply:
This is a ported over comment from the code I copied.